dys_function/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2022 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Dysfunctional "Node" detection for the SAFE Network.
+
+// For quick_error
+#![recursion_limit = "256"]
+#![doc(
+    html_logo_url = "https://github.com/maidsafe/QA/raw/master/Images/maidsafe_logo.png",
+    html_favicon_url = "https://maidsafe.net/img/favicon.ico",
+    test(attr(deny(warnings)))
+)]
+// Forbid some very bad patterns. Forbid is stronger than `deny`, preventing us from suppressing the
+// lint with `#[allow(...)]` et-all.
+#![forbid(
+    arithmetic_overflow,
+    mutable_transmutes,
+    no_mangle_const_items,
+    unknown_crate_types,
+    unsafe_code
+)]
+// Turn on some additional warnings to encourage good style.
+#![warn(
+    missing_debug_implementations,
+    missing_docs,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_extern_crates,
+    unused_import_braces,
+    unused_qualifications,
+    unused_results,
+    clippy::unicode_not_nfc
+)]
+
+#[macro_use]
+extern crate tracing;
+
+mod connections;
+mod detection;
+mod network_knowledge;
+mod operations;
+
+use xor_name::XorName;
+
+use dashmap::DashMap;
+use itertools::Itertools;
+use std::collections::BTreeSet;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+/// Some reproducible xorname derived from the operation. This is a permanent reference needed for logging all dysfunction.
+type NodeIdentifier = XorName;
+
+// re declaration here to not create circular dep w/ sn atm.
+// TODO: depend on types once that's extracted
+type OperationId = [u8; 32];
+
+#[derive(Clone, Debug)]
+/// Dysfunctional node tracking. Allows various potential issues to be tracked and weighted,
+/// with unresposive or suspect nodes being noted on request, against which action can then be taken.
+pub struct DysfunctionDetection {
+    neighbour_count: usize,
+    communication_issues: Arc<DashMap<NodeIdentifier, usize>>, // count of comm issues
+    knowledge_issues: Arc<DashMap<NodeIdentifier, usize>>,     // count of comm issues
+    unfulfilled_ops: Arc<DashMap<NodeIdentifier, Arc<RwLock<Vec<OperationId>>>>>, // OperationId = [u8; 32]
```

Comment:
We could replace the count with the (local) timestamp of when the dysfunction occurred. (You can read unix timestamp as u64 like so: https://doc.rust-lang.org/std/time/struct.SystemTime.html#examples)

```suggestion
    communication_issues: Arc<DashMap<NodeIdentifier, Arc<RwLock<VecDeque<u64>>>>>, // count of comm issues
    knowledge_issues: Arc<DashMap<NodeIdentifier, Arc<RwLock<VecDeque<u64>>>>>,     // count of knowledge issues
```

That way we can dequeue old comm/knowledge issues, i.e. remove issues older than 10 minutes ago. (Also, fault detection logic starts to look more uniform across unfilled_ops and comm/knowledge_issues)